ovirt_imageio/client/_ovirt.py

```diff
@@ -0,0 +1,378 @@
+        shallow=shallow)
+
+    if isinstance(image, types.Disk):
+        transfer.disk = image
+    elif isinstance(image, types.DiskSnapshot):
+        transfer.snapshot = image
+    else:
+        raise ValueError(f"Cannot transfer {Repr(image)}")
+
+    transfers_service = con.system_service().image_transfers_service()
+    transfer = transfers_service.add(transfer)
+
+    # At this point the transfer owns the disk and will delete the disk
+    # if the transfer is canceled, or if finalizing the transfer fails.
+
+    transfer_service = _transfer_service(con, transfer.id)
+
+    while True:
+        try:
+            transfer = transfer_service.get()
+        except sdk.NotFoundError:
+            # The system has removed the disk and the transfer.
+            raise RuntimeError(f"Transfer {transfer.id} was removed")
+
+        if transfer.phase == types.ImageTransferPhase.FINISHED_FAILURE:
+            # The system will remove the disk and the transfer soon.
+            raise RuntimeError(f"Transfer {transfer.id} has failed")
+
+        if transfer.phase == types.ImageTransferPhase.PAUSED_SYSTEM:
+            transfer_service.cancel()
+            raise RuntimeError(f"Transfer {transfer.id} was paused by system")
+
+        if transfer.phase == types.ImageTransferPhase.TRANSFERRING:
+            break
+
+        if transfer.phase != types.ImageTransferPhase.INITIALIZING:
+            transfer_service.cancel()
+            raise RuntimeError(
+                f"Unexpected transfer {transfer.id} phase {transfer.phase}")
+
+        if time.monotonic() > deadline:
+            log.info("Cancelling transfer %s", transfer.id)
+            transfer_service.cancel()
+            raise RuntimeError(
+                f"Timed out waiting for transfer {transfer.id}")
+
+        time.sleep(1)
+
+    # Log the transfer host name. This is very useful for
+    # troubleshooting.
+    transfer.host = con.follow_link(transfer.host)
+
+    log.info("Transfer %r ready on host %r in %.1f seconds",
+             transfer.id, transfer.host.name, time.monotonic() - start)
+
+    return transfer
+
+
+def cancel_transfer(con, transfer):
+    """
+    Cancel a transfer and remove the disk for upload transfer.
+
+    There is not need to cancel a download transfer, it can always be
+    finalized.
+    """
+    log.info("Cancelling transfer %r", transfer.id)
+    transfer_service = _transfer_service(con, transfer.id)
+    transfer_service.cancel()
+
+
+def finalize_transfer(con, transfer, disk, timeout=300):
+    """
+    Finalize a transfer, making the transfer disk available.
+
+    If finalizing succeeds: the disk status will change to OK and
+    transfer's phase will change to FINISHED_SUCCESS.
+    On upload errors: the disk status will change to ILLEGAL, transfer's
+    phase will change to FINISHED_FAILURE and the disk will be removed.
+    In both cases the transfer entity:
+     a. prior to 4.4.7: removed shortly after the command finishes
```

Comment:
Sounds better with "is" before the removed.

Reply:
Agree, will update in the next version.

Reply:
Fixed in current version.